atc/metric/emitter/influxdb.go

```diff
@@ -46,48 +58,83 @@ func (config *InfluxDBConfig) NewEmitter() (metric.Emitter, error) {
 	}
 
 	return &InfluxDBEmitter{
-		client:   client,
-		database: config.Database,
+		Client:   	client,
+		Database: 	config.Database,
+		BatchSize: 	int(config.BatchSize),
+		BatchDuration: 	config.BatchDuration,
 	}, nil
 }
 
-func (emitter *InfluxDBEmitter) Emit(logger lager.Logger, event metric.Event) {
+func emitBatch(emitter *InfluxDBEmitter, logger lager.Logger, events []metric.Event) {
+
+	logger.Debug("influxdb-emit-batch", lager.Data{
+		"size": len(events),
+	})
 	bp, err := influxclient.NewBatchPoints(influxclient.BatchPointsConfig{
-		Database: emitter.database,
+		Database: emitter.Database,
 	})
 	if err != nil {
 		logger.Error("failed-to-construct-batch-points", err)
 		return
 	}
 
-	tags := map[string]string{
-		"host": event.Host,
-	}
-
-	for k, v := range event.Attributes {
-		tags[k] = v
+	for _, event := range events {
+		tags := map[string]string{
+			"host": event.Host,
+		}
+
+		for k, v := range event.Attributes {
+			tags[k] = v
+		}
+
+		point, err := influxclient.NewPoint(
+			event.Name,
+			tags,
+			map[string]interface{}{
+				"value": event.Value,
+				"state": string(event.State),
+			},
+			event.Time,
+		)
+		if err != nil {
+			logger.Error("failed-to-construct-point", err)
+			continue
+		}
+
+		bp.AddPoint(point)
 	}
 
-	point, err := influxclient.NewPoint(
-		event.Name,
-		tags,
-		map[string]interface{}{
-			"value": event.Value,
-			"state": string(event.State),
-		},
-		event.Time,
-	)
+	err = emitter.Client.Write(bp)
 	if err != nil {
-		logger.Error("failed-to-construct-point", err)
+		logger.Error("failed-to-send-points",
+			errors.Wrap(metric.ErrFailedToEmit, err.Error()))
 		return
 	}
+	logger.Info("influxdb-emitter-fork-influxdb-batch-emitted", lager.Data{
+		"size": len(events),
+	})
+}
 
-	bp.AddPoint(point)
 
-	err = emitter.client.Write(bp)
-	if err != nil {
-		logger.Error("failed-to-send-points",
-			errors.Wrap(metric.ErrFailedToEmit, err.Error()))
-		return
+func (emitter *InfluxDBEmitter) Emit(logger lager.Logger, event metric.Event) {
+	batch = append(batch, event)
+	duration := time.Since(lastBatchTime)
+	if len(batch) >= emitter.BatchSize || duration >= emitter.BatchDuration {
+		logger.Debug("influxdb-pre-emit-batch", lager.Data{
+			"influxdb-batch-size": emitter.BatchSize,
+			"current-batch-size": len(batch),
+			"influxdb-batch-duration": emitter.BatchDuration,
+			"current-duration": duration,
+		})
+		emitter.SubmitBatch(logger)
 	}
 }
+
+func (emitter *InfluxDBEmitter) SubmitBatch(logger lager.Logger) {
```

Comment:
Just checking, is this thread-safe? I guess it is because all of the emits come from an emit loop? :thinking:

Reply:
@vito, yes, that is my understanding. The below chunk of code in emit.go will serialize the emission of events to the InfluxDB code. emitLoop -> emitter.Emit -> emitter.SubmitBatch

func emitLoop() {
	for emission := range emissions {
		emitter.Emit(emission.logger.Session("emit"), emission.event)
	}
}